bitacross-worker/bitacross/core/bc-relayer-registry/src/lib.rs

```diff
@@ -0,0 +1,197 @@
+		}
+	}
+
+	impl SealedIO for RelayerRegistrySeal {
+		type Error = Error;
+		type Unsealed = RelayerRegistryMap;
+
+		fn unseal(&self) -> Result<Self::Unsealed> {
+			Ok(unseal(self.path()).map(|b| Decode::decode(&mut b.as_slice()))??)
+		}
+
+		fn seal(&self, unsealed: &Self::Unsealed) -> Result<()> {
+			info!("Seal relayer registry to file: {:?}", unsealed);
+			Ok(unsealed.using_encoded(|bytes| seal(bytes, self.path()))?)
+		}
+	}
+}
+
+#[cfg(feature = "sgx")]
+use sgx::*;
+
+pub trait RelayerRegistryUpdater {
+	fn init(&self) -> RegistryResult<()>;
+	fn update(&self, account: Identity) -> RegistryResult<()>;
+	fn remove(&self, account: Identity) -> RegistryResult<()>;
+	fn contains_key(&self, account: Identity) -> bool;
+}
+
+// TODO: unit-test
+impl RelayerRegistryUpdater for RelayerRegistry {
+	#[cfg(feature = "std")]
+	fn init(&self) -> RegistryResult<()> {
+		Ok(())
+	}
+
+	#[cfg(feature = "std")]
+	fn update(&self, _account: Identity) -> RegistryResult<()> {
+		Ok(())
+	}
+
+	#[cfg(feature = "std")]
+	fn remove(&self, _account: Identity) -> RegistryResult<()> {
+		Ok(())
+	}
+
+	#[cfg(feature = "std")]
+	fn contains_key(&self, _account: Identity) -> bool {
+		true
+	}
+
+	// if `RELAYER_REGISTRY_FILE` exists, unseal and init from it
+	// otherwise create a new instance and seal to static file
+	#[cfg(feature = "sgx")]
+	fn init(&self) -> RegistryResult<()> {
+		let enclave_seal = RelayerRegistrySeal::new(self.seal_path.clone());
+		if SgxFile::open(RELAYER_REGISTRY_FILE).is_err() {
+			info!(
+				"[Enclave] RelayerRegistry file not found, creating new! {}",
+				RELAYER_REGISTRY_FILE
+			);
+			let mut registry = GLOBAL_RELAYER_REGISTRY
+				.registry
+				.write()
+				.map_err(|_| RegistryError::PoisonLock)?;
+			enclave_seal.seal(&*registry)
+		} else {
+			let m = enclave_seal.unseal()?;
+			info!("[Enclave] RelayerRegistry unsealed from file: {:?}", m);
+			let mut registry = GLOBAL_RELAYER_REGISTRY
+				.registry
+				.write()
+				.map_err(|_| RegistryError::PoisonLock)?;
+			*registry = m;
+			Ok(())
+		}
+	}
+
+	#[cfg(feature = "sgx")]
+	fn update(&self, account: Identity) -> RegistryResult<()> {
+		let mut registry = GLOBAL_RELAYER_REGISTRY
+			.registry
+			.write()
+			.map_err(|_| RegistryError::PoisonLock)?;
+		registry.insert(account, ());
+		RelayerRegistrySeal::new(self.seal_path.clone()).seal(&*registry)
+	}
+
+	#[cfg(feature = "sgx")]
+	fn remove(&self, account: Identity) -> RegistryResult<()> {
+		let mut registry = GLOBAL_RELAYER_REGISTRY
+			.registry
+			.write()
+			.map_err(|_| RegistryError::PoisonLock)?;
+		let old_value = registry.remove(&account);
+		if old_value.is_some() {
+			return RelayerRegistrySeal::new(self.seal_path.clone()).seal(&*registry)
+		}
+		Ok(())
+	}
+
+	#[cfg(feature = "sgx")]
+	fn contains_key(&self, account: Identity) -> bool {
+		// Using unwrap becaused poisoned locks are unrecoverable errors
+		let mut registry = GLOBAL_RELAYER_REGISTRY.registry.write().unwrap();
```

Comment:
Can it be `read()` instead of `write()` ?

Reply:
Using `read()` now, Performed a test